File: Miscellaneous_Functions/Spare_Table/sparse_table.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <numeric>
#include <string>
#include <cctype>

using namespace std;

class SparseTable {
private:
    vector<vector<int>> lookup;           //Sparse table
    vector<int> bin_log;                   
    int n;
    string func_name;

    int operation(int x, int y) {
        switch (func_name[0]) { 
            case 'm': // min, max
                if (func_name == "min") return min(x, y);
                if (func_name == "max") return max(x, y);
                break;
            case 'g': // gcd
                if (func_name == "gcd") return gcd(x, y);
                break;
            case 'l': // lcm
                if (func_name == "lcm") return lcm(x, y);
                break;
            case 'a': // and
                if (func_name == "and") return x & y;
                break;
            case 'o': // or
                if (func_name == "or") return x | y;
                break;
        }
        throw invalid_argument("Unsupported function name: " + func_name);
    }

    int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b); 
    }

    int lcm(int a, int b) {
        if (a == 0 || b == 0) return 0;
        return abs(a * b) / gcd(a, b);
    }

public:
    void build(vector<int>& arr, int size, string func) {
        transform(func.begin(), func.end(), func.begin(), ::tolower);
        func_name = func;
        n = size;

        bin_log.resize(n + 1);
        bin_log[1] = 0;
        for (int i = 2; i <= n; i++) {
            bin_log[i] = bin_log[i / 2] + 1;
        }

        int max_log = bin_log[n];
        lookup.assign(n, vector<int>(max_log + 1));           // Base case: single elements

        for (int i = 0; i < n; i++) {
            lookup[i][0] = arr[i];
        }

        for (int k = 1; k <= max_log; k++) {
            for (int i = 0; i + (1 << k) <= n; i++) {
                lookup[i][k] = operation(lookup[i][k - 1], lookup[i + (1 << (k - 1))][k - 1]);
            }
        }
    }

    int query(int L, int R) {
        int k = bin_log[R - L + 1];
        return operation(lookup[L][k], lookup[R - (1 << k) + 1][k]);
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++)
        cin >> arr[i];

    string func_name;
    cin >> func_name;

    SparseTable st;
    st.build(arr, n, func_name);

    int q;
    cin >> q;
    while (q--) {
        int L, R;
        cin >> L >> R;
        cout << st.query(L, R) << "\n";
    }
    return 0;
}


/*int main() {
    vector<int> arr = {7, 8, 9, 4, 5, 6, 7, 4, 1};
    int n = arr.size();
    string func_name = "min";

    SparseTable st;
    st.build(arr, n, func_name);

    vector<pair<int, int>> queries = {{0, 4}, {4, 7}, {7, 8}};
    cout << "Function: " << func_name << endl;
    for (auto [L, R] : queries) {
        cout << "Query (" << L << ", " << R << ") -> " << st.query(L, R) << endl;
    }

    return 0;
}*/

File: Miscellaneous_Functions/Spare_Table/sparse_table.java
import java.util.*;

class SparseTable {
    private int[][] lookup;
    private int[] binLog;
    private int n;
    private String funcName;

    private int operation(int x, int y) {
        switch (funcName) {
            case "min":
                return Math.min(x, y);
            case "max":
                return Math.max(x, y);
            case "gcd":
                return gcd(x, y);
            case "lcm":
                return lcm(x, y);
            case "and":
                return x & y;
            case "or":
                return x | y;
            default:
                throw new IllegalArgumentException("Unsupported function name: " + funcName);
        }
    }

    private int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }

    private int lcm(int a, int b) {
        if (a == 0 || b == 0) return 0;
        return Math.abs(a * b) / gcd(a, b);
    }

    public void build(int[] arr, int size, String func) {
        funcName = func.toLowerCase();
        n = size;

        binLog = new int[n + 1];
        binLog[1] = 0;
        for (int i = 2; i <= n; i++) {
            binLog[i] = binLog[i / 2] + 1;
        }

        int maxLog = binLog[n];
        lookup = new int[n][maxLog + 1];

        for (int i = 0; i < n; i++) {
            lookup[i][0] = arr[i];
        }

        for (int k = 1; k <= maxLog; k++) {
            for (int i = 0; i + (1 << k) <= n; i++) {
                lookup[i][k] = operation(lookup[i][k - 1], lookup[i + (1 << (k - 1))][k - 1]);
            }
        }
    }

    public int query(int L, int R) {
        int k = binLog[R - L + 1];
        return operation(lookup[L][k], lookup[R - (1 << k) + 1][k]);
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        String funcName = sc.next();

        SparseTable st = new SparseTable();
        st.build(arr, n, funcName);

        int q = sc.nextInt();
        while (q-- > 0) {
            int L = sc.nextInt();
            int R = sc.nextInt();
            System.out.println(st.query(L, R));
        }

        sc.close();
    }
}
